### DISCUSSION TITLE
Audio input stream starts, but always fails to read with `AAUDIO_ERROR_INVALID_STATE`

### DISCUSSION THREAD

**Discussion by CatalinVoss** [google/oboe#1451] [Q&A] [answered]
I'm on Android target SDK 28, NDK version 21.4.7075529, oboe 1.6.1 linked via the out-of-the-box `find_package` setup.

I'm opening a vanilla input stream like so:

```
void SpeechService::setup_stream() {
    oboe::AudioStreamBuilder builder;
    builder.setDirection(oboe::Direction::Input);
    builder.setPerformanceMode(oboe::PerformanceMode::LowLatency);
    builder.setChannelCount(1);
    builder.setChannelConversionAllowed(true);
    builder.setFormat(oboe::AudioFormat::I16);
    builder.setSampleRate(kSampleRate);
    builder.setCallback(this);
    oboe::Result result = builder.openStream(input_stream);
    if (result != oboe::Result::OK){
        LOGE("Error opening audio stream: %d", (result));
    } else {
        LOGI("Successfully set up native oboe stream");
    }
}
```

After opening it, I can verify that the stream is in "started" state. I have audio recording permission.

Then my callback follows the sample code:

```
oboe::DataCallbackResult SpeechService::onAudioReady(oboe::AudioStream *oboe_stream,
                                                     void *audio_data,
                                                     int32_t num_frames) {
    const int64_t timeout_nanos = 0; // for a non-blocking read

    auto result = input_stream->read(audio_data, num_frames, timeout_nanos);
    ...
}
```

This gets called with a small number of frames -- 320 or 192 or so. However, the read call always fails with the error `ErrorInvalidState = -895, // AAUDIO_ERROR_INVALID_STATE`.

Any idea how I should approach debugging this?

**Update 1:** On another device, the same code crashes on the read line with error

```
A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0xc in tid 20237 (AudioRecord)
```

**Update 2:** If I call `builder.setSampleRateConversionQuality(oboe::SampleRateConversionQuality::Medium)` on the builder, the crash goes away and result returns OK, but `result.value()` in the read callback is always 0.

**Comment by philburk** [comment, marked as answer]
You opened the stream with a callback. So you do not need to read it. The
data is already passed to you in the audioData buffer.

**Reply by CatalinVoss** [reply]
Thank you so much! Seems like I clearly misunderstood the instructions here! (I think because the "full guide" example reads into an output stream – I now realize it's an output stream! – I assumed you'd always have to read manually.)

**Comment by philburk** [comment]
We should update the docs to make it more clear. That code is "reading" from a renderer into a buffer that is provided by the callback. Maybe we should just write some white noise into the buffer to make it more clear That's  the easiest thing to do in a couple lines of code.
